src/targets/gpu/kernels/include/migraphx/kernels/gqa_softmax.hpp

```diff
@@ -0,0 +1,155 @@
+/*
+ * The MIT License (MIT)
+ *
+ * Copyright (c) 2015-2024 Advanced Micro Devices, Inc. All rights reserved.
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.  IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+#ifndef MIGRAPHX_GUARD_KERNELS_GQA_SOFTMAX_HPP
+#define MIGRAPHX_GUARD_KERNELS_GQA_SOFTMAX_HPP
+
+#include <migraphx/kernels/group_query_attention.hpp>
+#include <migraphx/kernels/index.hpp>
+#include <migraphx/kernels/tensor_view.hpp>
+
+namespace migraphx {
+
+template <class T>
+__device__ bool float_equal(T x, T y)
+{
+    return isfinite(x) and isfinite(y) and nextafterf(x, numeric_lowest<T>()) <= y and
+           nextafterf(x, numeric_max<T>()) >= y;
+}
+
+template <class T>
+__device__ void softmax_inplace(T score, int n, int d)
+{
+    for(int j = 0; j < n; ++j)
+    {
+        auto x = score + j * d;
+        auto y = x;
+
+        // e^x is represented as infinity if x is large enough, like 100.f.
+        // Infinity divided by Infinity is a NAN. Thus, softmax gets a NAN if
+        // one or more item are large enough. a math transform as below is
+        // leveraged to get a stable softmax: e^xi/(e^x1 + ...e^xn) = e^(xi -
+        // max) / (e^(x1 - max) + ... + e^(xn - max))
+        float max = -numeric_max<float>();
+        for(int i = 0; i < d; i++)
+        {
+            if(max < x[i])
+                max = x[i];
+        }
+        for(int i = 0; i < d; i++)
+        {
+            y[i] = expf(x[i] - max);
+        }
+
+        float sum        = 0.0;
+        const float zero = 0.0;
+        for(int i = 0; i < d; i++)
+        {
+            sum += x[i];
+        }
+
+        if(float_equal(sum, zero))
+        {
+            for(int i = 0; i < d; i++)
+            {
+                y[i] = 1.0f / static_cast<float>(d);
```

Comment:
Odd, when can we ever get into this case? The sum of exponents should never become negative and never converge to zero even if x[i] ins negative since e^(x) is never negative and you're doing things inplace

Reply:
Removed